src/goto-instrument/contracts/contracts.cpp

```
@@ -879,16 +877,27 @@ void code_contractst::apply_function_contract(
   // Insert havoc instructions immediately before the call site.
   new_program.destructive_append(havoc_instructions);
 
-  // To remove the function call, insert statements related to the assumption.
-  // Then, replace the function call with a SKIP statement.
-  if(!ensures.is_false())
+  // Generate: assume(ensures)
+  for(auto &clause : instantiated_ensures_clauses)
   {
-    is_fresh.update_ensures(ensures_pair.first);
-    new_program.destructive_append(ensures_pair.first);
+    if(!clause.is_false())
+    {
+      source_locationt _location = clause.source_location();
+      _location.set_comment("Assume ensures clause");
```

maybe throw a hard error if the clause is statically false ?

---

See https://github.com/diffblue/cbmc/pull/7159#discussion_r984978712